cloud-service-providers/google-cloud/vertexai/python/nim-vertexai.ipynb

```diff
@@ -0,0 +1,1485 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "73d3d2c3-92ad-49c5-85a5-229602c25e06",
+   "metadata": {},
+   "source": [
+    "Initialize Vertex AI SDK for Python"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "uYMNxGozkZyj",
+   "metadata": {
+    "executionInfo": {
+     "elapsed": 1,
+     "status": "ok",
+     "timestamp": 1721251252748,
+     "user": {
+      "displayName": "",
+      "userId": ""
+     },
+     "user_tz": 240
+    },
+    "id": "uYMNxGozkZyj",
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "from google.cloud import aiplatform\n",
+    "\n",
+    "aiplatform.init(project=project_id, location=region, staging_bucket=bucket_url)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "43c98fc3-d8a2-4abd-bd86-b1e120f07c4a",
+   "metadata": {},
+   "source": [
+    "GCP Configuration"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "455fad6a-e213-4e7e-baab-84f78b17fbe4",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def run_bash_cmd(cmd):\n",
+    "    import subprocess\n",
+    "\n",
+    "    if isinstance(cmd, str):\n",
+    "        process = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, shell=True, text=True)\n",
+    "    elif isinstance(cmd, list):\n",
+    "        process = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, shell=False, text=True)\n",
+    "        \n",
+    "    output, error = process.communicate()\n",
+    "    if error:\n",
+    "        raise Exception(error)\n",
+    "    else:\n",
+    "        print(output)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "5d570257-69fd-48e3-a558-8d861db11818",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "bash_cmd = f\"\"\"\n",
+    "    export region={region}\n",
+    "    gcloud config set ai_platform/region {region}\n",
+    "    gcloud config set project {project_id}\n",
+    "    gcloud auth configure-docker {region}-docker.pkg.dev\n",
+    "    \"\"\"\n",
+    "run_bash_cmd(bash_cmd)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "4cb66464-949b-4998-b9fc-620acba7fd8a",
+   "metadata": {
+    "tags": []
+   },
+   "source": [
+    "### NIM Container\n",
+    "\n",
+    "* **NGC_API_KEY**\n",
+    "\n",
+    "There are two options of NIM container:\n",
+    "\n",
+    "1) If NIM is not airgapped, `NGC_API_KEY` is needed when launch NIM container. The credential will be used as an environment variable, and will show on Model Registry Version Details UI.\n",
+    "\n",
+    "2) If NIM is airgapped, `NGC_API_KEY` is defined in `docker-compose.yaml` during image creation, then not required when launch NIM container as it will be automatically authorized, and will not show on Model Registry UI. Additional development steps are required.\n",
```

Comment:
What is meant by "automatically authorized"? If the container is airgapped, will the model be predownloaded into the environment?

Reply:
"automatically authorized" means NGC key is no needed when run airgap NIM. Similar to the new NIM release download-to-cache function and purpose. Model files are manually downloaded with NGC CLI commands in the notebook, to create the airgap container. When run the airgap container after creation, model files are not needed to present in host machine.

Reply:
Please note this airgap workaround here created before official release of new NIM functions download-to-cache. They should serve similar purpose and methodology.